src/payment/strategies/masterpass/masterpass-payment-strategy.ts

```diff
@@ -0,0 +1,127 @@
+import {
+    InvalidArgumentError,
+    MissingDataError,
+    MissingDataErrorType
+} from '../../../common/error/errors';
+import { StoreConfig } from '../../../config';
+import { OrderActionCreator, OrderRequestBody } from '../../../order';
+import { PaymentArgumentInvalidError } from '../../errors';
+import PaymentMethod from '../../payment-method';
+
+import { MasterpassCheckoutCallback, MasterpassCheckoutOptions } from './masterpass';
+import MasterpassScriptLoader from './masterpass-script-loader';
+
+export default class MasterpassPaymentStrategy extends PaymentStrategy {
+    private _paymentMethod?: PaymentMethod;
+    private _paymentGateway?: string;
+
+    constructor(
+        store: CheckoutStore,
+        private _orderActionCreator: OrderActionCreator,
+        private _paymentActionCreator: PaymentActionCreator,
+        private _paymentMethodActionCreator: PaymentMethodActionCreator,
+        private _masterpassScriptLoader: MasterpassScriptLoader
+    ) {
+        super(store);
+    }
+
+    initialize(options: PaymentInitializeOptions): Promise<InternalCheckoutSelectors> {
+        const { methodId, masterpass: masterpassOptions } = options;
+
+        if (!masterpassOptions) {
+            throw new InvalidArgumentError('Unable to initialize payment because masterpass options is missing');
+        }
+
+        const state = this._store.getState();
+        this._paymentMethod = state.paymentMethods.getPaymentMethod(methodId);
+
+        return Promise.resolve(this._hasPaymentInfo()).then(hasPaymentInfo => {
+            if (!hasPaymentInfo) {
+                return this._masterpassClientSetup().then(checkoutCallback => checkoutCallback());
+            }
+
+            this._paymentGateway = masterpassOptions.gateway;
+
+            if (masterpassOptions.onPaymentSelect) {
+                masterpassOptions.onPaymentSelect();
+            }
+        }).then(() => {
+            return super.initialize(options);
+        });
+    }
+
+    deinitialize(options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        this._paymentMethod = undefined;
+        this._paymentGateway = undefined;
+
+        return super.deinitialize(options);
+    }
+
+    execute(payload: OrderRequestBody, options?: PaymentRequestOptions): Promise<InternalCheckoutSelectors> {
+        const { payment, ...order } = payload;
+
+        if (!(payment && payment.paymentData)) {
+            throw new PaymentArgumentInvalidError(['payment.paymentData']);
+        }
+
+        const { paymentData } = payment;
+        const methodId = this._paymentGateway;
+
+        if (!methodId) {
+            throw new MissingDataError(MissingDataErrorType.MissingPaymentMethod);
+        }
+
+        return this._store.dispatch(this._orderActionCreator.submitOrder(order, options))
+            .then(() => this._store.dispatch(this._paymentMethodActionCreator.loadPaymentMethod(methodId)))
```

Comment:
Is there any reason why you need to reload the payment method again?

Reply:
Here the payment method which uses the masterpass wallet is going to be loaded, i.e stripe

Reply:
Yeah, so why do you need to do that? Unless I'm missing something, it doesn't seem you're using the data at all? We want to save a network call if it's possible.

Reply:
oh, I see, you are swapping the `methodId` for the `gateway` and then using that in the `paymentActionCreator`, which will try to get the gateway as a payment provider. is that right?

I don't really know how are we making these `gateways` be valid payment methods but not to show in the `/payments` endpoint...

Reply:
@icatalina, @davidchin yes that's right from masterpass we only get the nonce (payment token) that will be used with the payment provider which has the wallet configured, for instance if the gateway is stripe, we load the stripe payment method in order to be able to do the submitPayment call using stripe.